src/PowerShellEditorServices/Language/TokenOperations.cs

```diff
@@ -0,0 +1,271 @@
+                }
+                if ((tokenStack.Count > 0) & (token.Kind == endTokenKind)) {
+                    result.Add(CreateFoldingReference(tokenStack.Pop(), token, matchKind));
+                }
+            }
+            return result;
+        }
+
+        /// <summary>
+        /// Given a array tokens finds a specific token
+        /// </summary>
+        static private List<FoldingReference> MatchTokenElement(
+            Token[] tokens,
+            TokenKind tokenKind,
+            string matchKind)
+        {
+            List<FoldingReference> result = new List<FoldingReference>();
+            foreach (Token token in tokens)
+            {
+                if ((token.Kind == tokenKind) & (token.Extent.StartLineNumber != token.Extent.EndLineNumber)) {
+                    result.Add(CreateFoldingReference(token, token, matchKind));
+                }
+            }
+            return result;
+        }
+
+        /// <summary>
+        /// Returns true if a Token is a block comment;
+        /// - Must be a TokenKind.comment
+        /// - Must be preceeded by TokenKind.NewLine
+        /// - Token text must start with a '#'.false  This is because comment regions
+        ///   start with '&lt;#' but have the same TokenKind
+        /// </summary>
+        static private bool IsBlockComment(int index, Token[] tokens) {
+            Token thisToken = tokens[index];
+            if (thisToken.Kind != TokenKind.Comment) { return false; }
+            if (index == 0) { return true; }
+            if (tokens[index - 1].Kind != TokenKind.NewLine) { return false; }
+            return thisToken.Text.StartsWith("#");
+        }
+
+        /// <summary>
+        /// Finding blocks of comment tokens is more complicated as the newline characters are not
+        /// classed as comments.  To workaround this we search for valid block comments (See IsBlockCmment)
+        /// and then determine contiguous line numbers from there
+        /// </summary>
+        static private List<FoldingReference> MatchBlockCommentTokenElement(
+            Token[] tokens,
+            string matchKind)
+        {
+            // This regular expressions is used to detect a line comment (as opposed to an inline comment), that is not a region
+            // block directive i.e.
+            // - No text between the beginning of the line and `#`
+            // - Comment does start with region
+            // - Comment does start with endregion
+            string lineCommentText = @"\s*#(?!region\b|endregion\b)";
+
+            List<FoldingReference> result = new List<FoldingReference>();
+            Token startToken = null;
+            int nextLine = -1;
+            for (int index = 0; index < tokens.Length; index++)
+            {
+                Token thisToken = tokens[index];
+                if ((IsBlockComment(index, tokens)) & (Regex.IsMatch(thisToken.Text, lineCommentText, RegexOptions.IgnoreCase))) {
```

Comment:
I noticed that you're using the Boolean logical AND operator. Can you just use `&&` instead? Then if x is false, it wont go to y.

```suggestion
                if ((IsBlockComment(index, tokens)) && (Regex.IsMatch(thisToken.Text, lineCommentText, RegexOptions.IgnoreCase))) {
```

Reply:
Fixed